x-pack/legacy/plugins/lens/public/indexpattern_plugin/state_helpers.ts

```diff
@@ -124,11 +124,10 @@ export function deleteColumn({
   layerId: string;
   columnId: string;
 }): IndexPatternPrivateState {
-  const newColumns = adjustColumnReferencesForChangedColumn(
-    state.layers[layerId].columns,
-    columnId
-  );
-  delete newColumns[columnId];
+  const hypotheticalColumns = { ...state.layers[layerId].columns };
```

Comment:
I think we'd benefit from some immutable lodash-like helpers. I've had to do a `dissoc` in other places (which is what you're doing here). It would be nice for this to read:

```js
const newColumns = adjustColumnReferencesForChangedColumn(
  dissoc(columnId, state.layers[layerId].columns),
  columnId
);
```

Also, I've noticed that we have a lot of nested accessing layered throughout our code, and I wonder if we'd benefit form a set of selector helper functions:

```js
// Instead of: state.layers[layerId].columns
layerColumns(state, layerId);
```

Not sure. I've used selectors a lot in previous React apps, and the nice thing is that if you ever change your state structure, you only really have to change your selectors, and the rest of the app pretty much works.

Reply:
100% agree about nested accessors being replaced by selectors.